### ISSUE TITLE
Feature request: Add an option to choose NCHW/NHWC in convolutional layers

### ISSUE THREAD

**Issue by jihunchoi** [pytorch/pytorch#1220]
As far as I know, when we use cudnn on convolution operations, there exists an option to specify whether an input data is in NCHW format or in NHWC format.
It seems that currently PyTorch only supports NCHW format, thus one has to apply transpose operation and then make the results contiguous explicitly.
I think it would be useful if I can directly use NHWC format for convolutional operations.

**Comment by soumith** [comment]
what's your use-case for NHWC? We made an explicit choice to simplify the interface to just the NCHW format

**Comment by jihunchoi** [comment]
Actually I'm trying to use convoluational operations to natural language processing application, and it is quite common that an input data has shape (batch, length, hidden_size) in NLP tasks.
If the overhead resulted by "transpose-contiguous" is negligible, I would use that kind of approach without worry, but I don't know about the overhead much.

**Comment by fmassa** [comment]
Hum, we actually have kernels for Conv1d working on `NxWxC`, but we have decided to simplify the interface.
I can't speak for NLP use-cases, but I'd vote for keeping a single data layout for 2d convolutions, as `NCHW` is going to be much faster thanks to better cudnn support , and adding a new data layout will just make using the library more complex and error prone.

**Comment by apaszke** [comment]
Is it a big problem to use a transpose of the data throughout the network? I talked about that with @jekbradbury and he said it doesn't limit NLP use cases in any way.

**Comment by jihunchoi** [comment]
You mean using the transposed data (batch, hidden_size, length) for the entire network instead of using (batch, length, hidden), right?
When only the convolutional layers are used then it would be just OK but if one wants to augment the network to use other modules, for example RNN or Linear, then it should still be transposed.
Of course I know transposing an input, apply the module, and then transposing back can be really easily implemented. There's no reason that I want to avoid that approach, if performance degradation is not severe.

**Comment by apaszke** [comment]
I think that the perf of NCHW + transpose might be similar to NHWC alone (it's a faster and more optimized data format for GPUs). But I guess we can leave the issue open with a very low priority, unless someone can demonstrate some benefits.

**Comment by ppwwyyxx** [comment]
From the related comments here: https://github.com/tensorflow/tensorflow/issues/15897#issuecomment-356725027, NCHW is not necessarily more efficient than NHWC in Volta and future architectures.

**Comment by soumith** [comment]
@jph00 we are thinking of solving this, but with graph rewriting + JIT. Exposing data layouts at the user level will start making things chaotic / ugly for the users (they need to start thinking about one extra type on which not all functions are implemented)

**Comment by ddkang** [comment]
While I agree that extra complexity is generally, shouldn't the user be _allowed_ to change things if they want? For example, it could not be exposed in any of the examples, but documented for those who want fast prototyping with performance.

**Comment by davidcpage** [comment]
CuDNN 7.4 is out with NHWC pooling and batchnorm so I believe that this is now crucial for getting the best CNN training performance on Volta (as explained here https://devblogs.nvidia.com/tensor-core-ai-performance-milestones/ .)
It would be great if this could be exposed before the graph rewriting stuff appears. I personally like the explicit approach that's been taken elsewhere in PyTorch and would be keen to see this applied to layout too.

**Comment by dzhulgakov** [comment]
Documenting discussion from Slack:

For long-term, it'd be connected with NamedTensor proposal. For the short-term, we prefer an option that doesn't change Conv API. Such option is "NCHW sizes, NHWC strides".

Pros: no change to the public API at this point. Op implementation can check similarly to `is_contiguous` something like `is_contiguous_nhwc(strides(), sizes())` which would verify corresponding product of dimensions.

Cons: some ops might accidentally reorder the tensors back to NCHW layout. Element-wise ops like relu should work fine though. In order to prevent it, we'd need to add additional flag on tensor but it's less clear at this moment how exactly do that.

@slayton58 was working on implementation of this approach

**Comment by VitalyFedyunin** [comment]
We are working on it #19092

**Comment by gchanan** [comment]
I think is more up-to-date, so I'm going to close this.

**Comment by MaratZakirov** [comment]
I have somehow related issue. I can not properly convert my model to TFlite due to it do not support NCHW format. Is there any reliable/simple way to convert my pytorch model to NHWC?
